core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushFilterThroughBoolOrAggregation.java

```diff
@@ -0,0 +1,296 @@
+                            .capturedAs(AGGREGATION)));
+        }
+
+        @Override
+        public Pattern<FilterNode> getPattern()
+        {
+            return pattern;
+        }
+
+        @Override
+        public Result apply(FilterNode node, Captures captures, Context context)
+        {
+            return pushFilter(node, captures.get(AGGREGATION), Optional.empty(), plannerContext, context);
+        }
+    }
+
+    @VisibleForTesting
+    public static final class PushFilterThroughBoolOrAggregationWithProject
+            implements Rule<FilterNode>
+    {
+        private static final Capture<ProjectNode> PROJECT = newCapture();
+        private static final Capture<AggregationNode> AGGREGATION = newCapture();
+
+        private final PlannerContext plannerContext;
+        private final Pattern<FilterNode> pattern;
+
+        public PushFilterThroughBoolOrAggregationWithProject(PlannerContext plannerContext)
+        {
+            this.plannerContext = requireNonNull(plannerContext, "plannerContext is null");
+            this.pattern = filter()
+                    .with(source().matching(project()
+                            .matching(ProjectNode::isIdentity)
+                            .capturedAs(PROJECT)
+                            .with(source().matching(aggregation()
+                                    .matching(PushFilterThroughBoolOrAggregation::isGroupedBoolOrWithMask)
+                                    .capturedAs(AGGREGATION)))));
+        }
+
+        @Override
+        public Pattern<FilterNode> getPattern()
+        {
+            return pattern;
+        }
+
+        @Override
+        public Result apply(FilterNode node, Captures captures, Context context)
+        {
+            return pushFilter(node, captures.get(AGGREGATION), Optional.of(captures.get(PROJECT)), plannerContext, context);
+        }
+    }
+
+    private static Result pushFilter(FilterNode filterNode, AggregationNode aggregationNode, Optional<ProjectNode> projectNode, PlannerContext plannerContext, Context context)
+    {
+        Symbol boolOrSymbol = getOnlyElement(aggregationNode.getAggregations().keySet());
+        Aggregation aggregation = getOnlyElement(aggregationNode.getAggregations().values());
+
+        DomainTranslator.ExtractionResult extractionResult = getExtractionResult(plannerContext, context.getSession(), filterNode.getPredicate());
+        TupleDomain<Symbol> tupleDomain = extractionResult.getTupleDomain();
+        Expression remainingExpression = extractionResult.getRemainingExpression();
+
+        if (tupleDomain.isNone()) {
+            // Filter predicate is never satisfied. Replace filter with empty values.
+            return Result.ofPlanNode(new ValuesNode(filterNode.getId(), filterNode.getOutputSymbols(), ImmutableList.of()));
+        }
+
+        Optional<Expression> boolSymbolInRemaining = Optional.empty();
```

Comment:
bool condition can only be in one place, either tuple domain or remaining expression. I think it will always be in `tupleDomain`, but please check that

Reply:
`CoalesceExpression` is placed into `remainingExpression` by `getExtractionResult`. And CoalesceExpression is a case for correlated exists.